Allow users to customize the footer of a login theme

thomasdarimont opened the pull request:
Previously, users had to copy the whole template.ftl file or resort to css/js hacks to add a custom footer.

This PR adds new footer templates to the `base/login` and `keycloak.v2/login` themes to allow users
to configure a footer for the login box.

The new `footer.ftl` template contains a content macro that can be added at the bottom of the "login box".

Adjusted `template.ftl` in base login theme to render the macro from `footer.ftl`

The new templates are empty by default to avoid conflicts with existing themes. 

Fixes #31390

<!---
Please read https://github.com/keycloak/keycloak/blob/main/CONTRIBUTING.md and follow these guidelines when contributing to Keycloak
-->

thomasdarimont committed:
Allow users to customize the footer of a login theme (#31390)

Previously, users had to copy the whole template.ftl file or resort to css/js hacks to
add a custom footer.

This PR adds new footer.ftl template to the base login theme to allow users
to configure a footer for the login box, as well as the page footer.

The new footer.ftl template contains a content macro that can be added at the bottom of the login page.

Adjusted template.ftl in base/login and keycloak.v2 theme to render footer.ftl macro.

Fixes #31390

Signed-off-by: Thomas Darimont <thomas.darimont@googlemail.com>

thomasdarimont commented:
An example custom login theme with a custom footer could look like this:

theme.properties:
```
parent=keycloak
import=common/keycloak
styles=css/login.css css/custom-login.css
```

resources/css/custom-login.css
```css
#kc-footer-links {
    text-align: center;
}
#kc-footer-links li{
    display:inline-block;
}
```

`footer.ftl` file:
```html
<#macro content>
    <!-- hello from footer.html -->
    <hr id="footer-divider">
    <ul id="kc-footer-links">
        <li><a href="https://apps.training.test/apps/site#home">Home</a></li>
        <li><a href="https://apps.training.test/apps/site#about">About</a></li>
        <li><a href="https://apps.training.test/apps/site#services">Services</a></li>
        <li><a href="https://apps.training.test/apps/site#contact">Contact</a></li>
    </ul>
</#macro>
```

Gives the following result:
![image](https://github.com/user-attachments/assets/b7edb900-fb45-4d57-bae8-7729e79de00d)

With Identity Providers:
![image](https://github.com/user-attachments/assets/b9596cb5-35e3-45f1-9f63-1f141698fa40)

This could also be added to the keycloak.v2 login theme.

cturkalj commented:
Great idea!
After looking at the result, would it be possible to also add a (page) footer at the bottom of the page (instead of the "login form")?
There you could add other things (like a copyright notice) without making the login form look overcrowded.

thomasdarimont commented:
@cturkalj I thought about that too. How about the following:

`login-footer.ftl` contains a macro that's rendered at the end of the "login box" (like my current footer.ftl above).
`login.ftl` contains a macro that's rendered at the end of the page as you suggested.

Both macros would be empty by default.

Would that work for you?

thomasdarimont commented:
@edewit thanks for your review! What do you think about the suggestions from @cturkalj?

cturkalj commented:
@thomasdarimont:
Perfect, thank you!

thomasdarimont commented:
I did a quick example for the suggested extension:
https://gist.github.com/thomasdarimont/8021ef4bf715aae2cb28ca2ca3151ab7#file-template-ftl-L169

jonkoops commented on themes/src/main/resources/theme/base/login/template.ftl:
```diff
@@ -163,7 +164,9 @@
         </div>
       </div>
-
+      <footer>
```
Since this element is empty, perhaps we can simply omit it the tag for now? If folks want to add a footer element they can put it in the footer template.

thomasdarimont replied:
Thanks for pointing that out @jonkoops , yes we can give the user complete control of the fragment then.
Would you mind taking a look at the gist in https://github.com/keycloak/keycloak/pull/31391#issuecomment-2235836319 ?

I updated the gist and included your suggestion.

thomasdarimont commented:
I updated the PR based on the suggestions. I now use empty footer.ftl and login-footer.ftl templates by default. 
Users have full control over the respective contents.

Allthough I think the footer.ftl will rarely be used as the login-footer.ftl is sufficient for many scenarios already. Users can use the footer.ftl to include javascript add the very bottom of the page if necessary.

jonkoops reviewed (dismissed):
LGTM!

thomasdarimont commented:
Added note to release notes

ahus1 commented:
@thomasdarimont - there is a conflict with the release notes. Please rebase and resolve.

thomasdarimont commented:
@ahus1 Thanks for the review. I rebased the PR and resolved the conflict.

I also added support for the custom footer(s) to the login.v2 theme which I learned will become the default login theme from now on.

However, there are some issues with the login theme v2 which I wrote on the maintainers mailing list. So I'd recommend to wait for feedback from the maintainers before merging this PR.

thomasdarimont commented:
I also tested this with the keycloak.v2 login theme.

Theme minimal.v2/login, see: https://gist.github.com/thomasdarimont/5fe725319a5fc3c8242669deddd934a8

Custom theme based on Login V2 with Footer 
![image](https://github.com/user-attachments/assets/3135635a-02b8-46ae-af5b-91ea4c8be54e)

Custom theme based on Login V2 with Identity Providers and Footer 
![image](https://github.com/user-attachments/assets/10e913dd-adf9-4745-8498-54e5214e5258)

thomasdarimont commented:
@cturkalj I experimented a bit further with the footer customizations and I think it would be sufficient to only have one `footer.ftl` file that we add at the bottom of the login box.  Links that should be shown at the bottom of the page can be moved there via CSS. This would make the templating easier, as the current login page styling of the keycloak and keycloak.v2 theme does not support adding links outside of the "login box" anyways.

Can you think of something that would not be possible if we only had one footer.ftl ( the login-footer.ftl in the current examples)?

jonkoops commented on docs/documentation/release_notes/topics/26_0_0.adoc:
```diff
@@ -50,6 +50,14 @@ The `java-keystore` key provider, which allows loading a realm key from an exter
 
 For more information about this subject, see link:{adminguide_link}#realm_keys[Configuring realm keys].
 
+= Customizable footer in login Theme
```
Can we also add this information to the theming documentation?

thomasdarimont replied:
Yes!

thomasdarimont replied:
Just updated the docs, please check :)

cturkalj commented:
@thomasdarimont:
I was able to use the login-footer.ftl only and have the footer at the bottom of the page with css (thanks for the hint). 
Besides that, I currently don't have in other use cases in mind.

Now that we have just one (login)-footer.ftl left, should it be renamed it back to footer.ftl?

thomasdarimont commented:
I revised the PR to only contain 1 additional template file -> `footer.ftl`

- Added section to the release-notes
- Added section to the upgrading guide
- Added section to the themes section in the server development docs

ahus1 reviewed (approved):
Approving based on @jonkoops' and @edewit's review. Thanks!
